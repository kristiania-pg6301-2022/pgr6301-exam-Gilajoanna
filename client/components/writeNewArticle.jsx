import * as React from "react";
import { useContext, useState } from "react";
import { AppContext } from "../globals/AppContext";

function ArticleForm({ label, value, onChangeValue }) {
  return (
    <div>
      <label>{label}</label>{" "}
      <input
        type={"text"}
        name={"title"}
        value={value}
        onChange={(e) => onChangeValue(e.target.value)}
      />
    </div>
  );
}

export function WriteNewArticle() {
  const { postArticle } = useContext(AppContext);

  const [title, setTitle] = useState("");
  const [category, setCategory] = useState("");
  const [content, setContent] = useState("");
  const [author, setAuthor] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    if (!title || !category || !author || !content) {
      alert("One or more inputs are empty. Please try again.");
    }

    await postArticle({ title, category, author, content });

    setTitle("");
    setCategory("");
    setAuthor("");
    setContent("");
  }

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <h1>Write new article</h1>

        <ArticleForm label={"Title:"} value={title} onChangeValue={setTitle} />
        <ArticleForm
          label={"Category:"}
          value={category}
          onChangeValue={setCategory}
        />
        <ArticleForm
          label={"Content:"}
          value={content}
          onChangeValue={setContent}
        />
        <ArticleForm
          label={"Author:"}
          value={author}
          onChangeValue={setAuthor}
        />
        <button type={"submit"}>Publish</button>
      </form>
    </div>
  );
}
