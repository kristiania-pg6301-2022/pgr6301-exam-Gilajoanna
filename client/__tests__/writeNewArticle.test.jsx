import * as ReactDOM from "react-dom";
import * as React from "react";
import { WriteNewArticle } from "../writeNewArticle";
import { act } from "react-dom/test-utils";

describe("write new article", () => {
  it("shows new article form", async () => {
    const element = document.createElement("div");
    await act(async () => ReactDOM.render(<WriteNewArticle />, element));
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form label")).map((e) => e.innerHTML)
    ).toEqual(["Title:", "Category:", "Content:", "Author:"]);
  });
});
