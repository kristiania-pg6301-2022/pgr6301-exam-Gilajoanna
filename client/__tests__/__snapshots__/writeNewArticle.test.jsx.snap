// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Write new article shows new article form 1`] = `"<form><h1>Write new article</h1><div><label>Title:</label> <input type=\\"text\\" name=\\"title\\" value=\\"\\"></div><div><label>Category:</label> <input type=\\"text\\" name=\\"title\\" value=\\"\\"></div><div><label>Content:</label> <input type=\\"text\\" name=\\"title\\" value=\\"\\"></div><div><label>Author:</label> <input type=\\"text\\" name=\\"title\\" value=\\"\\"></div><button type=\\"submit\\">Publish</button></form>"`;
